llvm/projects/evm-sdk/evm-stdlib/include/storage.h

```diff
@@ -0,0 +1,256 @@
+  using ValueType = typename Impl::ValueType;
+  using Impl::Impl;
+
+  auto operator[](uint256_t k) {
+    auto k1 = __builtin_evm_sha3_vargs(Impl::slot(), k);
+    if constexpr (std::is_integral_v<ValueType>) {
+      return Slot<ValueType>(k1);
+    } else {
+      return ValueType(k1);
+    }
+  }
+
+  auto operator[](uint256_t k) const {
+    return const_cast<MapBase*>(this)->operator[](k);
+  }
+
+  void remove(uint256_t key) const {
+    auto slot = __builtin_evm_sha3_vargs(Impl::slot(), key);
+    details::clear<ValueType>(slot);
+  }
+};
+
+template<typename T>
+using Map = MapBase<SlotImpl<T>>;
+
+template<typename T, uint256_t Slot>
+using MapStatic = MapBase<SlotStaticImpl<T, Slot>>;
+
+/*******************************************************************************
+ * Iterable map.
+ * Unlike the Map this class allows to iterate over all elements in map, remove
+ * elements, find elements and get size of the map.
+ */
+template<typename Impl>
+class IterableMapBase: private Impl {
+public:
+  using Impl::Impl;
+
+  using ValueType = typename Impl::ValueType;
+
+  static constexpr unsigned FIELDS_NUM = 3;
+
+  unsigned size() const {
+    return Impl::keys.size();
+  }
+
+  bool empty() const {
+    return size() == 0;
+  }
+
+  bool contains(uint256_t key) const {
+    return Impl::inserted[key];
+  }
+
+  uint256_t get_key_at(uint256_t index) const {
+    require(index < size());
+    return Impl::keys[index].get();
+  }
+
+  void remove(uint256_t key) {
+    if (!contains(key)) {
+      return;
+    }
+    Impl::inserted[key] = false;
+    Impl::values.remove(key);
+    auto index = Impl::indexes[key].get();
+    auto last_key = Impl::keys[size() - 1].get();
+
+    Impl::indexes[last_key] = index;
+    Impl::indexes.remove(0);
+
+    Impl::keys[index] = last_key;
+    Impl::keys.pop();
+  }
+
+  auto operator[](uint256_t key) {
+    if (Impl::inserted[key].get()) {
+      return Impl::values[key];
+    } else {
+      Impl::inserted[key] = true;
+      Impl::indexes[key] = Impl::keys.size();
+      Impl::keys.push().set(key);
+      return Impl::values[key];
+    }
+  }
+};
+
+template<typename T>
+struct IterableMapImpl {
+  using ValueType = T;
+
+  IterableMapImpl(uint256_t slot): keys(slot), values(slot + 1), indexes(slot + 2), inserted(slot + 3) {}
+
+  static constexpr unsigned FIELDS_NUM = 4;
+
+  Vector<uint256_t> keys;
+  Map<ValueType> values;
+  Map<uint256_t> indexes;
+  Map<bool> inserted;
+};
+
+template<typename T, uint256_t Slot>
+struct IterableMapStaticImpl {
+  using ValueType = T;
+
+  static constexpr unsigned FIELDS_NUM = 4;
+
+  VectorStatic<uint256_t, Slot> keys;
+  MapStatic<ValueType, Slot + 1> values;
+  MapStatic<uint256_t, Slot + 2> indexes;
+  MapStatic<bool, Slot + 3> inserted;
+};
+
+template<typename T>
+using IterableMap = IterableMapBase<IterableMapImpl<T>>;
+
+template<typename T, uint256_t Slot>
+using IterableMapStatic = IterableMapBase<IterableMapStaticImpl<T, Slot>>;
```

Comment:
Same question as the static aray. Does it mean that we would be able to change the size in runtime still?

Reply:
Explanation is above.